// elysia/dist/compose.js
import{parse as e}from"fast-querystring";import{sign as r}from"cookie-signature";import{mapEarlyResponse as t,mapResponse as s,mapCompactResponse as n,isNotEmpty as o}from"./handler";import{NotFoundError as a,ValidationError as i,InternalServerError as c,ERROR_CODE as l}from"./error";import{parseCookie as f}from"./cookie";let u=new Headers().toJSON,d=RegExp(" (\\w+) = context","g"),p={value:0},m=({hasTrace:e,hasTraceSet:r=!1,hasTraceChildren:t=!1,addFn:s,condition:n={}})=>{if(!e)return()=>()=>{};{let e=t?"\nawait new Promise(r => {queueMicrotask(() => queueMicrotask(r))})\n":"\nawait new Promise(r => {queueMicrotask(r)})\n";return(t,{name:o,attribute:a="",unit:i=0}={})=>{let c=t.indexOf("."),l=-1===c;if("request"!==t&&"response"!==t&&!n[l?t:t.slice(0,c)])return()=>{r&&"afterHandle"===t&&s(e)};l?o||=t:o||="anonymous",s("\n"+`reporter.emit('event', { 
					id,
					event: '${t}',
					type: 'begin',
					name: '${o}',
					time: performance.now(),
					${l?`unit: ${i},`:""}
					${a}
				})`.replace(/(\t| |\n)/g,"")+"\n");let f=!1;return()=>{!f&&(f=!0,s("\n"+`reporter.emit('event', {
							id,
							event: '${t}',
							type: 'end',
							time: performance.now()
						})`.replace(/(\t| |\n)/g,"")+"\n"),r&&"afterHandle"===t&&s(e))}}}};export const hasReturn=e=>{let r=e.indexOf(")");return(// Is direct arrow function return eg. () => 1
61===e.charCodeAt(r+2)&&123!==e.charCodeAt(r+5)||e.includes("return"))};let h=(e,{injectResponse:r=""}={})=>({composeValidation:(r,t=`c.${r}`)=>e?`c.set.status = 400; throw new ValidationError(
'${r}',
${r},
${t}
)`:`c.set.status = 400; return new ValidationError(
	'${r}',
	${r},
	${t}
).toResponse(c.set.headers)`,composeResponseValidation:(t="r")=>{let s=e?`throw new ValidationError(
'response',
response[c.set.status],
${t}
)`:`return new ValidationError(
'response',
response[c.set.status],
${t}
).toResponse(c.set.headers)`;return`
${r}
		if(response[c.set.status]?.Check(${t}) === false) { 
	if(!(response instanceof Error))
		${s}
}
`}});export const isFnUse=(e,r)=>{r=(r=r.trimStart()).replaceAll(/^async /g,""),/^(\w+)\(/g.test(r)&&(r=r.slice(r.indexOf("(")));let t=// CharCode 40 is '('
40===r.charCodeAt(0)||r.startsWith("function")?r.slice(r.indexOf("(")+1,r.indexOf(")")):r.slice(0,r.indexOf("=")-1);if(""===t)return!1;let s=123===t.charCodeAt(0)?t.indexOf("..."):-1;// Using object destructuring
if(123===t.charCodeAt(0)){// Since Function already format the code, styling is enforced
if(t.includes(e))return!0;if(-1===s)return!1}// Match dot notation and named access
if(r.match(RegExp(`${t}(.${e}|\\["${e}"\\])`)))return!0;let n=-1!==s?t.slice(s+3,t.indexOf(" ",s+3)):void 0;if(r.match(RegExp(`${n}(.${e}|\\["${e}"\\])`)))return!0;let o=[t];for(let e of(n&&o.push(n),r.matchAll(d)))o.push(e[1]);let a=RegExp(`{.*?} = (${o.join("|")})`,"g");for(let[t]of r.matchAll(a))if(t.includes(`{ ${e}`)||t.includes(`, ${e}`))return!0;return!1};let y=e=>{e=(e=e.trimStart()).replaceAll(/^async /g,""),/^(\w+)\(/g.test(e)&&(e=e.slice(e.indexOf("(")));let r=// CharCode 40 is '('
40===e.charCodeAt(0)||e.startsWith("function")?e.slice(e.indexOf("(")+1,e.indexOf(")")):e.slice(0,e.indexOf("=")-1);if(""===r)return!1;let t=123===r.charCodeAt(0)?r.indexOf("..."):-1,s=-1!==t?r.slice(t+3,r.indexOf(" ",t+3)):void 0,n=[r];for(let r of(s&&n.push(s),e.matchAll(d)))n.push(r[1]);for(let r of n)if(RegExp(`\\b\\w+\\([^)]*\\b${r}\\b[^)]*\\)`).test(e))return!0;let o=RegExp(`{.*?} = (${n.join("|")})`,"g");for(let[r]of e.matchAll(o))if(RegExp(`\\b\\w+\\([^)]*\\b${r}\\b[^)]*\\)`).test(e))return!0;return!1},$=Symbol.for("TypeBox.Kind");export const hasType=(e,r)=>{if(r){if($ in r&&r[$]===e)return!0;if("object"===r.type){let t=r.properties;for(let r of Object.keys(t)){let s=t[r];if("object"===s.type){if(hasType(e,s))return!0}else if(s.anyOf){for(let r=0;r<s.anyOf.length;r++)if(hasType(e,s.anyOf[r]))return!0}if($ in s&&s[$]===e)return!0}return!1}return r.properties&&$ in r.properties&&r.properties[$]===e}};let g=Symbol.for("TypeBox.Transform");export const hasTransform=e=>{if(e){if("object"===e.type){let r=e.properties;for(let e of Object.keys(r)){let t=r[e];if("object"===t.type){if(hasTransform(t))return!0}else if(t.anyOf){for(let e=0;e<t.anyOf.length;e++)if(hasTransform(t.anyOf[e]))return!0}let s=g in t;if(s)return!0}return!1}return g in e||e.properties&&g in e.properties}};/**
 * This function will return the type of unioned if all unioned type is the same.
 * It's intent to use for content-type mapping only
 *
 * ```ts
 * t.Union([
 *   t.Object({
 *     password: t.String()
 *   }),
 *   t.Object({
 *     token: t.String()
 *   })
 * ])
 * ```
 */let b=e=>{if(!e)return;// @ts-ignore
let r=e?.schema;if(r&&"anyOf"in r){let e=!1,t=r.anyOf[0].type;for(let s of r.anyOf)if(s.type!==t){e=!0;break}if(!e)return t}// @ts-ignore
return e.schema?.type},k=/(?:return|=>) \S*\(/g;export const isAsync=e=>"AsyncFunction"===e.constructor.name||e.toString().match(k);export const composeHandler=({path:d,method:$,hooks:g,validator:k,handler:q,handleError:x,definitions:w,schema:H,onRequest:R,config:v,reporter:E})=>{let O=v.forceErrorEncapsulation||g.error.length>0||"undefined"==typeof Bun||g.onResponse.length>0||!!g.trace.length,A=g.onResponse.length?`
;(async () => {${g.onResponse.map((e,r)=>`await res${r}(c)`).join(";")}})();
`:"",C=g.trace.map(e=>e.toString()),F=!1;if(y(q.toString())&&(F=!0),!F){for(let[e,r]of Object.entries(g))if(Array.isArray(r)&&r.length&&["parse","transform","beforeHandle","afterHandle","onResponse"].includes(e)){for(let e of r)if("function"==typeof e&&y(e.toString())){F=!0;break}if(F)break}}let j={parse:C.some(e=>isFnUse("parse",e)),transform:C.some(e=>isFnUse("transform",e)),handle:C.some(e=>isFnUse("handle",e)),beforeHandle:C.some(e=>isFnUse("beforeHandle",e)),afterHandle:C.some(e=>isFnUse("afterHandle",e)),error:O||C.some(e=>isFnUse("error",e))},T=g.trace.length>0,S="";T&&(S+="\nconst id = c.$$requestId\n"),S+=O?"try {\n":"";let U=k||"GET"!==$&&"HEAD"!==$?[q,...g.transform,...g.beforeHandle,...g.afterHandle].map(e=>e.toString()):[],D=F||"GET"!==$&&"HEAD"!==$&&"none"!==g.type&&(!!k.body||!!g.type||U.some(e=>isFnUse("body",e))),V=F||k.headers||U.some(e=>isFnUse("headers",e)),N=F||k.cookie||U.some(e=>isFnUse("cookie",e)),_=k?.cookie?.schema,I="";if(_?.sign){if(!_.secrets)throw Error(`t.Cookie required secret which is not set in (${$}) ${d}.`);let e=_.secrets?"string"==typeof _.secrets?_.secrets:_.secrets[0]:void 0;if(I+=`const _setCookie = c.set.cookie
		if(_setCookie) {`,!0===_.sign)// encodeCookie += `if(_setCookie['${name}']?.value) { c.set.cookie['${name}'].value = signCookie(_setCookie['${name}'].value, '${secret}') }\n`
I+=`for(const [key, cookie] of Object.entries(_setCookie)) {
				c.set.cookie[key].value = signCookie(cookie.value, '${e}')
			}`;else for(let r of _.sign)// if (!(name in cookieMeta.properties)) continue
I+=`if(_setCookie['${r}']?.value) { c.set.cookie['${r}'].value = signCookie(_setCookie['${r}'].value, '${e}') }
`;I+="}\n"}let{composeValidation:B,composeResponseValidation:L}=h(O);if(V&&// This function is Bun specific
// @ts-ignore
(S+=u?`c.headers = c.request.headers.toJSON()
`:`c.headers = {}
                for (const [key, value] of c.request.headers.entries())
					c.headers[key] = value
				`),N){let e=(e,r)=>{// @ts-ignore
let t=_?.[e]??r;return t?"string"==typeof t?`${e}: '${t}',`:t instanceof Date?`${e}: new Date(${t.getTime()}),`:`${e}: ${t},`:""},r=_?`{
			secret: ${void 0!==_.secrets?"string"==typeof _.secrets?`'${_.secrets}'`:"["+_.secrets.reduce((e,r)=>e+`'${r}',`,"")+"]":"undefined"},
			sign: ${!0===_.sign||(void 0!==_.sign?"["+_.sign.reduce((e,r)=>e+`'${r}',`,"")+"]":"undefined")},
			${e("domain")}
			${e("expires")}
			${e("httpOnly")}
			${e("maxAge")}
			${e("path","/")}
			${e("priority")}
			${e("sameSite")}
			${e("secure")}
		}`:"undefined";V?S+=`
c.cookie = parseCookie(c.set, c.headers.cookie, ${r})
`:S+=`
c.cookie = parseCookie(c.set, c.request.headers.get('cookie'), ${r})
`}let P=F||k.query||U.some(e=>isFnUse("query",e));P&&(S+=`const url = c.request.url

		if(c.qi !== -1) {
			c.query ??= parseQuery(url.substring(c.qi + 1))
		} else {
			c.query ??= {}
		}
		`);let Q=g.trace.map(e=>e.toString()),W=Q.some(e=>isFnUse("set",e)||y(e)),G=W&&Q.some(e=>e.includes("children")||y(e));F||g.trace.some(e=>isFnUse("set",e.toString()));let K=W||N||U.some(e=>isFnUse("set",e))||R.some(e=>isFnUse("set",e.toString())),M=m({hasTrace:T,hasTraceSet:W,hasTraceChildren:G,condition:j,addFn:e=>{S+=e}}),J=D||W||isAsync(q)||g.parse.length>0||g.afterHandle.some(isAsync)||g.beforeHandle.some(isAsync)||g.transform.some(isAsync),z=M("parse",{unit:g.parse.length});if(D){let e=b(k?.body);if(g.type&&!Array.isArray(g.type)){if(g.type)switch(g.type){case"json":case"application/json":S+=`c.body = await c.request.json()
`;break;case"text":case"text/plain":S+=`c.body = await c.request.text()
`;break;case"urlencoded":case"application/x-www-form-urlencoded":S+=`c.body = parseQuery(await c.request.text())
`;break;case"arrayBuffer":case"application/octet-stream":S+=`c.body = await c.request.arrayBuffer()
`;break;case"formdata":case"multipart/form-data":S+=`c.body = {}

						const form = await c.request.formData()
						for (const key of form.keys()) {
							if (c.body[key])
								continue

							const value = form.getAll(key)
							if (value.length === 1)
								c.body[key] = value[0]
							else c.body[key] = value
						}
`}g.parse.length&&(S+="}}")}else{let r=(()=>{if(g.parse.length&&e&&!Array.isArray(g.type)){// @ts-ignore
let r=k?.body?.schema;if("object"===e&&(hasType("File",r)||hasType("Files",r)))return`c.body = {}
		
								const form = await c.request.formData()
								for (const key of form.keys()) {
									if (c.body[key])
										continue
			
									const value = form.getAll(key)
									if (value.length === 1)
										c.body[key] = value[0]
									else c.body[key] = value
								}`}})();if(r)S+=r;else{if(S+="\n"+(V?"let contentType = c.headers['content-type']":"let contentType = c.request.headers.get('content-type')")+`
				if (contentType) {
					const index = contentType.indexOf(';')
					if (index !== -1) contentType = contentType.substring(0, index)
`,g.parse.length){S+=`let used = false
`;let e=M("parse",{unit:g.parse.length});for(let e=0;e<g.parse.length;e++){let r=M("parse.unit",{name:g.parse[e].name}),t=`bo${e}`;0!==e&&(S+=`if(!used) {
`),S+=`let ${t} = parse[${e}](c, contentType)
if(${t} instanceof Promise) ${t} = await ${t}
if(${t} !== undefined) { c.body = ${t}; used = true }
`,r(),0!==e&&(S+="}")}e()}g.parse.length&&(S+="if (!used)"),S+=`
				switch (contentType) {
					case 'application/json':
						c.body = await c.request.json()
						break
				
					case 'text/plain':
						c.body = await c.request.text()
						break
				
					case 'application/x-www-form-urlencoded':
						c.body = parseQuery(await c.request.text())
						break
				
					case 'application/octet-stream':
						c.body = await c.request.arrayBuffer();
						break
				
					case 'multipart/form-data':
						c.body = {}
				
						const form = await c.request.formData()
						for (const key of form.keys()) {
							if (c.body[key])
								continue
				
							const value = form.getAll(key)
							if (value.length === 1)
								c.body[key] = value[0]
							else c.body[key] = value
						}
				
						break
					}
}
`}}S+="\n"}if(z(),g?.transform){let e=M("transform",{unit:g.transform.length});for(let e=0;e<g.transform.length;e++){let r=g.transform[e],t=M("transform.unit",{name:r.name});"derive"===r.$elysia?S+=isAsync(g.transform[e])?`Object.assign(c, await transform[${e}](c));`:`Object.assign(c, transform[${e}](c));`:S+=isAsync(g.transform[e])?`await transform[${e}](c);`:`transform[${e}](c);`,t()}e()}if(k&&(S+="\n",k.headers&&(S+=`if(headers.Check(c.headers) === false) {
				${B("headers")}
			}`,hasTransform(k.headers.schema)&&(S+=`
c.headers = headers.Decode(c.headers)
`)),k.params&&(S+=`if(params.Check(c.params) === false) {
				${B("params")}
			}`,hasTransform(k.params.schema)&&(S+=`
c.params = params.Decode(c.params)
`)),k.query&&(S+=`if(query.Check(c.query) === false) {
				${B("query")} 
			}`,hasTransform(k.query.schema)&&// Decode doesn't work with Object.create(null)
(S+=`
c.query = query.Decode(Object.assign({}, c.query))
`)),k.body&&(S+=`if(body.Check(c.body) === false) { 
				${B("body")}
			}`,hasTransform(k.body.schema)&&(S+=`
c.body = body.Decode(c.body)
`)),o(k.cookie?.schema.properties??{})&&(S+=`const cookieValue = {}
			for(const [key, value] of Object.entries(c.cookie))
				cookieValue[key] = value.value

			if(cookie.Check(cookieValue) === false) {
				${B("cookie","cookieValue")}
			}`,hasTransform(k.cookie.schema)&&(S+=`
c.cookie = params.Decode(c.cookie)
`))),g?.beforeHandle){let e=M("beforeHandle",{unit:g.beforeHandle.length});for(let e=0;e<g.beforeHandle.length;e++){let r=M("beforeHandle.unit",{name:g.beforeHandle[e].name}),t=`be${e}`,s=hasReturn(g.beforeHandle[e].toString());if(s){S+=isAsync(g.beforeHandle[e])?`let ${t} = await beforeHandle[${e}](c);
`:`let ${t} = beforeHandle[${e}](c);
`,r(),S+=`if(${t} !== undefined) {
`;let s=M("afterHandle",{unit:g.transform.length});if(g.afterHandle)for(let e=0;e<g.afterHandle.length;e++){let r=hasReturn(g.afterHandle[e].toString()),s=M("afterHandle.unit",{name:g.afterHandle[e].name});if(S+=`c.response = ${t}
`,r){let r=`af${e}`;S+=(isAsync(g.afterHandle[e])?`const ${r} = await afterHandle[${e}](c);
`:`const ${r} = afterHandle[${e}](c);
`)+`if(${r} !== undefined) { c.response = ${t} = ${r} }
`}else S+=isAsync(g.afterHandle[e])?`await afterHandle[${e}](c, ${t});
`:`afterHandle[${e}](c, ${t});
`;s()}s(),k.response&&(S+=L(t)),S+=I+`return mapEarlyResponse(${t}, c.set)}
`}else S+=isAsync(g.beforeHandle[e])?`await beforeHandle[${e}](c);
`:`beforeHandle[${e}](c);
`,r()}e()}if(g?.afterHandle.length){let e=M("handle",{name:q.name});g.afterHandle.length?S+=isAsync(q)?`let r = c.response = await handler(c);
`:`let r = c.response = handler(c);
`:S+=isAsync(q)?`let r = await handler(c);
`:`let r = handler(c);
`,e();let r=M("afterHandle",{unit:g.afterHandle.length});for(let e=0;e<g.afterHandle.length;e++){let t=`af${e}`,s=hasReturn(g.afterHandle[e].toString()),n=M("afterHandle.unit",{name:g.afterHandle[e].name});s?(k.response?S+=isAsync(g.afterHandle[e])?`let ${t} = await afterHandle[${e}](c)
`:`let ${t} = afterHandle[${e}](c)
`:S+=isAsync(g.afterHandle[e])?`let ${t} = mapEarlyResponse(await afterHandle[${e}](c), c.set)
`:`let ${t} = mapEarlyResponse(afterHandle[${e}](c), c.set)
`,n(),k.response?(S+=`if(${t} !== undefined) {`+L(t)+`${t} = mapEarlyResponse(${t}, c.set)
`+`if(${t}) {`,r(),S+=`return ${t} } }`):(S+=`if(${t}) {`,r(),S+=`return ${t}}
`)):(S+=isAsync(g.afterHandle[e])?`await afterHandle[${e}](c)
`:`afterHandle[${e}](c)
`,n())}r(),S+=`r = c.response
`,k.response&&(S+=L()),S+=I,K?S+=`return mapResponse(r, c.set)
`:S+=`return mapCompactResponse(r)
`}else{let e=M("handle",{name:q.name});if(k.response)S+=isAsync(q)?`const r = await handler(c);
`:`const r = handler(c);
`,e(),S+=L(),M("afterHandle")(),S+=I,K?S+=`return mapResponse(r, c.set)
`:S+=`return mapCompactResponse(r)
`;else if(j.handle||N)S+=isAsync(q)?`let r = await handler(c);
`:`let r = handler(c);
`,e(),M("afterHandle")(),S+=I,K?S+=`return mapResponse(r, c.set)
`:S+=`return mapCompactResponse(r)
`;else{e();let r=isAsync(q)?"await handler(c) ":"handler(c)";M("afterHandle")(),K?S+=`return mapResponse(${r}, c.set)
`:S+=`return mapCompactResponse(${r})
`}}if(O||A){S+=`
} catch(error) {`,J||(S+="return (async () => {"),S+=`const set = c.set

		if (!set.status || set.status < 300) set.status = 500
	`;let e=M("error",{unit:g.error.length});if(g.error.length)for(let e=0;e<g.error.length;e++){let r=`er${e}`,t=M("error.unit",{name:g.error[e].name});S+=`
let ${r} = handleErrors[${e}](
					Object.assign(c, {
						error: error,
						code: error.code ?? error[ERROR_CODE] ?? "UNKNOWN"
					})
				)
`,isAsync(g.error[e])&&(S+=`if (${r} instanceof Promise) ${r} = await ${r}
`),t(),S+=`${r} = mapEarlyResponse(${r}, set)
if (${r}) {return ${r} }
`}if(e(),S+=`return handleError(c, error)

`,J||(S+="})()"),S+="}",A||T){S+=" finally { ";let e=M("response",{unit:g.onResponse.length});S+=A,e(),S+="}"}}// console.log(fnLiteral)
S=`const { 
		handler,
		handleError,
		hooks: {
			transform,
			beforeHandle,
			afterHandle,
			parse,
			error: handleErrors,
			onResponse
		},
		validator: {
			body,
			headers,
			params,
			query,
			response,
			cookie
		},
		utils: {
			mapResponse,
			mapCompactResponse,
			mapEarlyResponse,
			parseQuery
		},
		error: {
			NotFoundError,
			ValidationError,
			InternalServerError
		},
		schema,
		definitions,
		ERROR_CODE,
		reporter,
		requestId,
		parseCookie,
		signCookie
	} = hooks

	${g.onResponse.length?`const ${g.onResponse.map((e,r)=>`res${r} = onResponse[${r}]`).join(",")}`:""}

	return ${J?"async":""} function(c) {
		${H&&w?"c.schema = schema; c.defs = definitions;":""}
		${S}
	}`;let X=Function("hooks",S);return X({handler:q,hooks:g,validator:k,handleError:x,utils:{mapResponse:s,mapCompactResponse:n,mapEarlyResponse:t,parseQuery:e},error:{NotFoundError:a,ValidationError:i,InternalServerError:c},schema:H,definitions:w,ERROR_CODE:l,reporter:E,requestId:p,parseCookie:f,signCookie:r})};export const composeGeneralHandler=e=>{let r="",s="";// @ts-ignore
for(let t of Object.keys(e.decorators))r+=`,${t}: app.decorators.${t}`;// @ts-ignore
let{router:n,staticRouter:o}=e,i=e.event.trace.length>0,c=`
	const route = find(request.method, path) ${n.root.ALL?'?? find("ALL", path)':""}
	if (route === null)
		return ${e.event.error.length?"app.handleError(ctx, notFound)":`new Response(error404, {
					status: 404
				})`}

	ctx.params = route.params

	return route.store(ctx)`,l="";for(let[e,{code:r,all:t}]of Object.entries(o.map))l+=`case '${e}':
switch(request.method) {
${r}
${t??"default: break map"}}

`;let f=e.event.request.some(isAsync);s+=`const {
		app,
		app: { store, router, staticRouter },
		mapEarlyResponse,
		NotFoundError,
		requestId,
		reporter
	} = data

	const notFound = new NotFoundError()

	${e.event.request.length?"const onRequest = app.event.request":""}

	${o.variables}

	const find = router.find.bind(router)
	const handleError = app.handleError.bind(this)

	${e.event.error.length?"":"const error404 = notFound.message.toString()"}

	return ${f?"async":""} function map(request) {
	`;let u=e.event.trace.map(e=>e.toString()),d=m({hasTrace:i,hasTraceChildren:i&&u.some(e=>e.includes("children")||y(e)),condition:{request:u.some(e=>isFnUse("request",e)||y(e))},addFn:e=>{s+=e}});if(e.event.request.length){s+=`
			${i?"const id = +requestId.value++":""}

			const ctx = {
				request,
				store,
				set: {
					cookie: {},
					headers: {},
					status: 200
				}
				${i?",$$requestId: +id":""}
				${r}
			}
		`;let t=d("request",{attribute:"ctx",unit:e.event.request.length});s+=`try {
`;for(let r=0;r<e.event.request.length;r++){let t=e.event.request[r],n=hasReturn(t.toString()),o=isAsync(t),a=d("request.unit",{name:e.event.request[r].name}),i=`re${r}`;n?(s+=`const ${i} = mapEarlyResponse(
					${o?"await":""} onRequest[${r}](ctx),
					ctx.set
				)
`,a(),s+=`if(${i}) return ${i}
`):(s+=`${o?"await":""} onRequest[${r}](ctx)
`,a())}s+=`} catch (error) {
			return app.handleError(ctx, error)
		}`,t(),s+=`
		const url = request.url,
		s = url.indexOf('/', 11),
		i = ctx.qi = url.indexOf('?', s + 1),
		path = ctx.path = i === -1 ? url.substring(s) : url.substring(s, i);`}else s+=`
		const url = request.url,
			s = url.indexOf('/', 11),
			qi = url.indexOf('?', s + 1),
			path = qi === -1
				? url.substring(s)
				: url.substring(s, qi)

		${i?"const id = +requestId.value++":""}

		const ctx = {
			request,
			store,
			qi,
			path,
			set: {
				headers: {},
				status: 200
			}
			${i?",$$requestId: id":""}
			${r}
		}`,d("request",{unit:e.event.request.length,attribute:u.some(e=>isFnUse("context",e))||u.some(e=>isFnUse("store",e))||u.some(e=>isFnUse("set",e))?"ctx":""})();return s+=`
		map: switch(path) {
			${l}

			default:
				break
		}

		${c}
	}`,// @ts-ignore
e.handleError=composeErrorHandler(e),Function("data",s)({app:e,mapEarlyResponse:t,NotFoundError:a,// @ts-ignore
reporter:e.reporter,requestId:p})};export const composeErrorHandler=e=>{let r=`const {
		app: { event: { error: onError, onResponse: res } },
		mapResponse,
		ERROR_CODE
	} = inject

	return ${e.event.error.find(isAsync)?"async":""} function(context, error) {
		const { set } = context
		`;for(let t=0;t<e.event.error.length;t++){let s=e.event.error[t],n=`${isAsync(s)?"await ":""}onError[${t}](
			Object.assign(context, {
				code: error.code ?? error[ERROR_CODE] ?? 'UNKNOWN',
				error
			})
		)`;hasReturn(s.toString())?r+=`const r${t} = ${n}; if(r${t} !== undefined) return mapResponse(r${t}, set)
`:r+=n+"\n"}return Function("inject",r+=`if(error.constructor.name === "ValidationError") {
		set.status = error.status ?? 400
		return new Response(
			error.message, 
			{ headers: set.headers, status: set.status }
		)
	} else {
		return new Response(error.message, { headers: set.headers, status: error.status ?? 500 })
	}
}`)({app:e,mapResponse:s,ERROR_CODE:l})};
